flocker/node/script.py

```diff
@@ -281,143 +320,312 @@ def get_service(self, reactor, options):
+    # cluster_uuid.  It probably should so that it can make sure it
+    # only talks to other nodes in the same cluster (maybe the
+    # authentication layer would mostly handle this but maybe not if
+    # you're slightly careless with credentials - also ZFS backend
+    # doesn't use TLS yet).
+    BackendDescription(
+        name=u"zfs", needs_reactor=True, needs_cluster_id=False,
+        api_factory=_zfs_storagepool, deployer_type=DeployerType.p2p,
+    ),
+    BackendDescription(
+        name=u"loopback", needs_reactor=False, needs_cluster_id=False,
+        # XXX compute_instance_id is the wrong type
+        api_factory=LoopbackBlockDeviceAPI.from_path,
+        deployer_type=DeployerType.block,
+    ),
+
+    # BackendDescription(
+    #     name=u"openstack", needs_reactor=False, needs_cluster_id=True,
+    #     factory=cinder_from_configuration,
+    #     deployer_type=DeployerType.block,
+    # ),
+]
 
-@implementer(ICommandLineVolumeScript)
-class GenericAgentScript(PRecord):
-    """
-    Implement top-level logic for the ``flocker-dataset-agent`` script.
+_DEFAULT_DEPLOYERS = {
+    DeployerType.p2p: lambda api, **kw:
+        P2PManifestationDeployer(volume_service=api, **kw),
+    DeployerType.block: lambda api, **kw:
+        BlockDeviceDeployer(block_device_api=api, **kw),
+}
+
+
+class AgentService(PRecord):
+    """
+    :ivar backends: ``BackendDescription`` instances describing how to use each
+        available storage backend.
+    :ivar deployers: Factories to create ``IDeployer`` providers given an API
+        object and some extra keyword arguments.  Keyed on a value from
+        ``DeployerType``.
+    :ivar node_credential: Credentials with which to configure this agent.
+    :ivar ca_certificate: The root certificate to use to validate the control
+        service certificate.
+    :ivar backend_name: The name of the storage driver to instantiate.  This
+        must name one of the items in ``backends``.
+    :ivar api_args: Extra arguments to pass to the factory from ``backends``.
+    """
+    backends = field(
+        factory=pvector, initial=_DEFAULT_BACKENDS, mandatory=True,
+    )
+    deployers = field(factory=pmap, initial=_DEFAULT_DEPLOYERS, mandatory=True)
+    reactor = field(initial=reactor, mandatory=True)
 
-    This is a temporary script, until the volume service can be created in
-    ``zfs_dataset_deployer``. The majority of this script will be in
-    ``flocker_dataset_agent_main`` and ``AgentScript``. See FLOC-1924.
-    """
-    def main(self, reactor, options, volume_service):
-        agent_config = options[u'agent-config']
-        configuration = yaml.safe_load(agent_config.getContent())
+    get_external_ip = field(initial=_get_external_ip, mandatory=True)
+
+    control_service_host = field(type=bytes, mandatory=True)
+    control_service_port = field(type=int, mandatory=True)
+
+    # Cannot use type=NodeCredential because one of the tests really wants to
```

Comment:
Mention in that particular test that if it is ever changed / removed, this could be improved.